From 52843c4173e68d4ba7ddefe51d417660ac8f14c1 Mon Sep 17 00:00:00 2001
From: Mingliang Liu <liuml07@apache.org>
Date: Wed, 30 Nov 2016 13:01:02 -0800
Subject: [PATCH 2178/2848] HADOOP-13830. Intermittent failure of
 ITestS3NContractRootDir#testRecursiveRootListing:
 "Can not create a Path from an empty string".
 Contributed by Steve Loughran

(cherry picked from commit 3fd844b99fdfae6be6e5e261f371d175aad14229)

Change-Id: I75955c1c02fd742cffcd83a365d46f22a5c171d9
---
 .../hadoop/fs/s3native/NativeS3FileSystem.java     |    7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3native/NativeS3FileSystem.java b/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3native/NativeS3FileSystem.java
index 9ae1061..e4e6037 100644
--- a/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3native/NativeS3FileSystem.java
+++ b/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3native/NativeS3FileSystem.java
@@ -574,7 +574,12 @@ else if (relativePath.endsWith(FOLDER_SUFFIX)) {
       for (String commonPrefix : listing.getCommonPrefixes()) {
         Path subpath = keyToPath(commonPrefix);
         String relativePath = pathUri.relativize(subpath.toUri()).getPath();
-        status.add(newDirectory(new Path(absolutePath, relativePath)));
+        // sometimes the common prefix includes the base dir (HADOOP-13830).
+        // avoid that problem by detecting it and keeping it out
+        // of the list
+        if (!relativePath.isEmpty()) {
+          status.add(newDirectory(new Path(absolutePath, relativePath)));
+        }
       }
       priorLastKey = listing.getPriorLastKey();
     } while (priorLastKey != null);
-- 
1.7.9.5

